In [25]:
import os
import json
import pandas as pd

# --- Local (with a .env file) ---
from dotenv import load_dotenv
load_dotenv()
API_KEY = os.environ["GROQ_API_KEY"]



# TODO: set API_KEY using ONE of the methods above.

# OpenAI-compatible client (works for Groq and OpenAI; Gemini users see their docs):
from openai import OpenAI

client = OpenAI(
    api_key=API_KEY,
    base_url="https://api.groq.com/openai/v1",  
)
MODEL = "llama-3.3-70b-versatile"                # or your provider's model name

print("Client ready.")

Client ready.


In [20]:
def ask_llm(user_prompt, system_prompt="You are a helpful assistant.",
            temperature=0.7, max_tokens=500):
     response = client.chat.completions.create(
         model=MODEL,
         messages=[
             {"role": "system", "content": system_prompt},
             {"role": "user",   "content": user_prompt},
         ],
         temperature=temperature,
         max_tokens=max_tokens,
     )
     return response

# TODO: Call it once with a simple question and print the answer.
response = ask_llm("What is family therapy?")
print(response.choices[0].message.content)

# TODO: Print response.usage as well — how many tokens did your call consume?
print(response.usage)

     

Family therapy, also known as family counseling or family systems therapy, is a type of psychotherapy that involves working with individuals and their family members to address relationship issues, conflicts, and communication problems. The goal of family therapy is to improve the overall functioning of the family unit, promote healthy relationships, and help family members develop more effective ways of interacting with each other.

In family therapy, a trained therapist works with the entire family, or a subset of family members, to:

1. Identify and address underlying patterns and dynamics that contribute to conflicts and problems.
2. Improve communication and problem-solving skills.
3. Enhance emotional expression and empathy among family members.
4. Develop more effective coping strategies and conflict resolution techniques.
5. Strengthen family bonds and promote a sense of unity and cooperation.

Family therapy can be beneficial for a wide range of issues, including:

* Relations

**Difference between system and user**

**system**:tells the AI its overall instructions, behavior, or rules for how it should respond.

**user**: Contains the actual question or request from the person using the AI

**What is Token**
A token is a small piece of text that an LLM processes. A token can be a whole word, part of a word, punctuation, or a space-related piece.

**Why do API providers bill per token rather than per request?**
API providers bill based on tokens because the amount of text the model processes varies greatly between requests. A short request might use only a few tokens, while a long conversation or document could use thousands or millions.

Therefore, charging per token is more accurate and fair than charging simply per request, because two requests can require vastly different amounts of computation.

In [21]:
question = "what is dark matter?"

# Temperature = 0.0
print("===== Temperature 0.0 =====")

for i in range(5):
    answer = ask_llm(
        question,
        temperature=0.0
    )
    print(f"{i + 1}. {answer}")


# Temperature = 1.2
print("\n===== Temperature 1.2 =====")

for i in range(5):
    answer = ask_llm(
        question,
        temperature=1.2
    )
    print(f"{i + 1}. {answer}")

===== Temperature 0.0 =====
1. ChatCompletion(id='chatcmpl-81354fcd-4a57-41a3-9c5c-c9eb55b64721', choices=[Choice(finish_reason='length', index=0, logprobs=None, message=ChatCompletionMessage(content='Dark matter is a mysterious and invisible form of matter that is thought to make up approximately 27% of the universe\'s total mass-energy density. It is called "dark" because it does not emit, absorb, or reflect any electromagnetic radiation, making it invisible to our telescopes.\n\nThe existence of dark matter was first proposed by Swiss astrophysicist Fritz Zwicky in the 1930s, based on his observations of the Coma galaxy cluster. He realized that the galaxies within the cluster were moving at much higher velocities than expected, suggesting that there was a large amount of unseen mass holding them together.\n\nSince then, a wealth of observational evidence has confirmed the existence of dark matter. Some of the key lines of evidence include:\n\n1. **Galactic Rotation Curves**: The ro

**At temperature 0.0**, the responses were generally very similar or sometimes identical. The model was more consistent and predictable.

**At temperature 1.2**, the responses were more varied and creative. The model produced different suggestions and wording across the five attempts.

**For the loan decision-support system** a low temperature, preferably around 0.0–0.3, is more appropriate. Loan decisions require consistency, predictability, and reliability rather than creative or random responses. Using a higher temperature could cause the system to give different recommendations for similar inputs, which would be undesirable for a financial decision-support system.

In [22]:
LETTERS = {
"L001": """Dear Sir/Madam,
My name is Akosua Mensah and I have been selling provisions at Makola Market for 12 years.
I am applying for a loan of GHS 8,000 to buy a deep freezer and expand into frozen foods.
My current stall makes about GHS 900 profit each month. I have saved GHS 2,500 with your
susu scheme over the past two years and I have never missed a contribution. I can repay
GHS 450 monthly over 20 months. My sister, a teacher, will stand as my guarantor.
Thank you for considering my application.""",

"L002": """Hello,
I am Kwame Boateng, a commercial driver in Kumasi. I need GHS 25,000 urgently to repair my
trotro engine and settle some personal debts. Business has been slow but it will surely
pick up after the festive season. I can pay back whenever the money comes. I do not have
collateral at the moment but God willing everything will be fine. Please help me quickly.""",

"L003": """Dear Loan Committee,
I am Efua Darko, owner of Darko Fashions, a registered dressmaking business in Takoradi
(registration no. BN-2019-4482). I employ three apprentices. I request GHS 15,000 to
purchase two industrial sewing machines and fabric stock ahead of the Christmas season.
Last year my December revenue alone was GHS 22,000; monthly profit averages GHS 2,800.
I hold a fixed deposit of GHS 5,000 with GCB which I can pledge. Proposed repayment:
GHS 1,100 monthly for 15 months. Attached are my sales records for the past 18 months.""",

"L004": """Good day,
My name is Yaw Owusu. I want a loan for my poultry farm at Nsawam. The amount is GHS 12,000
for feed and 500 new layers. I started the farm last year. Sometimes I make good money,
around GHS 1,500 in a good month, but bird flu affected us in March and I lost many birds.
I am rebuilding now. I can repay in 18 months. My uncle has agreed to guarantee the loan
with his taxi.""",

"L005": """Dear Manager,
I am writing on behalf of the Adenta Women's Weaving Cooperative (14 members). We seek
GHS 30,000 to buy a bulk order of yarn directly from the factory, cutting out middlemen and
raising our margins from 15% to about 35%. The cooperative has operated for 6 years and
holds GHS 9,000 in our group account. We propose repayment of GHS 2,000 monthly over
16 months, backed by our group savings and joint liability agreement.""",

"L006": """Hi,
This is Kofi. I saw your advert. I want GHS 50,000 to start a car washing business, a
provision shop, and also import phones from Dubai. I am 22 and full of energy. I have not
started any of these yet but my friends say I am very business minded. I will pay back in
one year when the businesses are booming. No collateral but I am trustworthy.""",
}

# Gold-standard labels for three letters (for Section 4 evaluation):
GOLD = {
  "L001": {"applicant_name": "Akosua Mensah", "amount_ghs": 8000,  "purpose": "buy deep freezer / expand into frozen foods",
           "monthly_profit_ghs": 900,  "has_collateral_or_guarantor": True,  "repayment_months": 20},
  "L003": {"applicant_name": "Efua Darko",    "amount_ghs": 15000, "purpose": "industrial sewing machines and fabric stock",
           "monthly_profit_ghs": 2800, "has_collateral_or_guarantor": True,  "repayment_months": 15},
  "L006": {"applicant_name": "Kofi",          "amount_ghs": 50000, "purpose": "car wash, provision shop, phone imports",
           "monthly_profit_ghs": None, "has_collateral_or_guarantor": False, "repayment_months": 12},
}

print(f"{len(LETTERS)} letters loaded.")

6 letters loaded.


In [23]:

# SUMMARY PROMPT V1 — NAIVE APPROACH

SUMMARY_PROMPT_V1 = "Summarize this:"


# Run V1 on L002
response_v1_l002 = ask_llm(
    SUMMARY_PROMPT_V1 + "\n\n" + LETTERS["L002"],
    temperature=0.7
)

v1_l002 = response_v1_l002.choices[0].message.content

print("===== V1 - L002 =====")
print(v1_l002)


# Run V1 on L006
response_v1_l006 = ask_llm(
    SUMMARY_PROMPT_V1 + "\n\n" + LETTERS["L006"],
    temperature=0.7
)

v1_l006 = response_v1_l006.choices[0].message.content

print("\n===== V1 - L006 =====")
print(v1_l006)


SUMMARY_SYSTEM_V2 = """
You are an assistant to a microfinance loan officer.

Summarize loan applications accurately and neutrally.

Follow these rules:
- Use only information provided in the application.
- Do not invent or assume any details.
- Include the applicant's name, loan amount, purpose,
  financial situation, collateral or guarantor, and
  repayment plan when available.
- Keep the summary factual and neutral.
- Write 3-4 sentences.
"""


SUMMARY_PROMPT_V2 = """
Summarize this loan application:

{letter_text}
"""


# Run V2 on L002
response_v2_l002 = ask_llm(
    SUMMARY_PROMPT_V2.format(letter_text=LETTERS["L002"]),
    system_prompt=SUMMARY_SYSTEM_V2,
    temperature=0.0
)

v2_l002 = response_v2_l002.choices[0].message.content

print("\n===== V2 - L002 =====")
print(v2_l002)


# Run V2 on L006
response_v2_l006 = ask_llm(
    SUMMARY_PROMPT_V2.format(letter_text=LETTERS["L006"]),
    system_prompt=SUMMARY_SYSTEM_V2,
    temperature=0.0
)

v2_l006 = response_v2_l006.choices[0].message.content

print("\n===== V2 - L006 =====")
print(v2_l006)


# SIDE-BY-SIDE COMPARISON

print("\n" + "=" * 70)
print("V1 VS V2 COMPARISON")
print("=" * 70)

print("\n--- L002 ---")

print("\nV1 (Naive Prompt):")
print(v1_l002)

print("\nV2 (Improved Prompt):")
print(v2_l002)


print("\n" + "-" * 70)

print("\n--- L006 ---")

print("\nV1 (Naive Prompt):")
print(v1_l006)

print("\nV2 (Improved Prompt):")
print(v2_l006)

===== V1 - L002 =====
Kwame Boateng, a commercial driver in Kumasi, is requesting GHS 25,000 to repair his vehicle's engine and pay off personal debts. He's experiencing a slow business period but expects it to improve after the festive season. He doesn't have collateral but promises to repay the loan as soon as possible.

===== V1 - L006 =====
Kofi, a 22-year-old, is seeking a loan of GHS 50,000 to start three businesses: a car washing service, a provision shop, and a phone import business from Dubai. He has no prior experience, but claims to be "business-minded" based on his friends' opinions. He promises to repay the loan within a year, once his businesses are successful, and offers his trustworthiness as assurance, since he has no collateral to provide.

===== V2 - L002 =====
Kwame Boateng is applying for a loan of GHS 25,000 to repair his trotro engine and settle personal debts. As a commercial driver in Kumasi, he is currently experiencing a slow business period but expects an im

**What concrete problems did V1's output have that V2 fixed?**

V1 was generally accurate, but it introduced some interpretations that were not directly stated in the applications and sometimes changed the wording in ways that could affect a loan officer's understanding.

For L002, V1 says:

**“is willing to repay the loan as soon as possible”**
However, the applicant actually said “I can pay back whenever the money comes.” These are not exactly the same. “As soon as possible” suggests a stronger or more definite willingness to repay than the original statement. 
V2 improved this by describing the repayment as a flexible repayment plan, which is closer to the applicant's actual wording.

For L006, V1 says:

**“He has no experience or collateral”**

The original letter only states that Kofi has not started any of the businesses yet. It does not explicitly say that he has no business experience. Therefore, “no experience” is an unsupported inference. 

V2 removed this claim and instead correctly stated:**“Kofi ... has not yet initiated any of these business ventures.”**

V1 also says:

**“promises to repay the loan within a year when his businesses are successful”**
while V2 more carefully states that the applicant proposes repayment “in one year, once the businesses are established.” This is more neutral and closer to what was actually provided.
Overall, V2 reduced unsupported interpretations and made the summaries more factual and neutral, which is important for a loan officer who may use the summary when assessing an applicant.

**Why is "no invented details" essential? What is this failure mode called?**
The instruction is essential because a loan decision-support system should base its analysis only on information actually provided by the applicant. If the LLM invents or assumes facts, it could incorrectly make an applicant appear more or less financially reliable and potentially influence a real lending decision.

This failure mode is commonly called an LLM hallucination (or hallucination). It occurs when a model generates information that is unsupported by, or not present in, the source material

In [ ]:

EXTRACT_PROMPT = """
Extract information from the loan application and return ONLY a valid JSON object.

The JSON object MUST contain EXACTLY these six keys:

{
  "applicant_name": "string",
  "amount_ghs": number,
  "purpose": "string",
  "monthly_profit_ghs": number or null,
  "has_collateral_or_guarantor": true or false,
  "repayment_months": number or null
}

Rules:
1. Return ONLY the JSON object.
2. Do not include explanations, comments, or markdown.
3. Use EXACTLY the six keys shown above.
4. If a field is not stated in the letter, use null.
5. Do not guess, assume, or invent any information.
6. amount_ghs must be a number, not a string.
7. monthly_profit_ghs must be a number or null.
8. repayment_months must be a number or null.
9. has_collateral_or_guarantor must be true or false.
10. If either collateral OR a guarantor is mentioned, set
    has_collateral_or_guarantor to true.
11. Keep the purpose concise but faithful to the application.

ONE WORKED EXAMPLE:

Letter:
My name is Ama Mensah. I operate a small bakery in Tema and have been
in business for four years. I am requesting GHS 10,000 to purchase a
new oven and additional baking supplies. My average monthly profit is
GHS 1,500. My brother has agreed to act as my guarantor. I propose to
repay the loan over 10 months.

Correct JSON:
{
  "applicant_name": "Ama Mensah",
  "amount_ghs": 10000,
  "purpose": "purchase a new oven and additional baking supplies",
  "monthly_profit_ghs": 1500,
  "has_collateral_or_guarantor": true,
  "repayment_months": 10
}

Now extract the fields from this loan application:

LETTER_TEXT_HERE
"""



# EXTRACTION FUNCTION


def extract_fields(letter_text):

    # Use replace() instead of .format()
    # This prevents JSON braces from causing KeyError
    prompt = EXTRACT_PROMPT.replace(
        "LETTER_TEXT_HERE",
        letter_text
    )

    try:

        response = ask_llm(
            prompt,
            temperature=0.0
        )

        # Get the model's actual text response
        result = response.choices[0].message.content.strip()

        # Print the raw response if needed for debugging
        # print("RAW RESPONSE:", result)

        
        # Remove ```json fences
        

        if result.startswith("```json"):
            result = result[7:].strip()

        elif result.startswith("```"):
            result = result[3:].strip()

        if result.endswith("```"):
            result = result[:-3].strip()

      
        # Parse JSON
       

        data = json.loads(result)

       
        # Make sure response is a dictionary
       

        if not isinstance(data, dict):
            print("WARNING: Response is not a JSON object.")
            return None

       
        # Required keys
       

        required_keys = {
            "applicant_name",
            "amount_ghs",
            "purpose",
            "monthly_profit_ghs",
            "has_collateral_or_guarantor",
            "repayment_months"
        }

        # Check that exactly the required keys exist
        if set(data.keys()) != required_keys:
            print("WARNING: JSON does not contain exactly the required keys.")
            print("Returned keys:", list(data.keys()))
            return None

        return data

    except json.JSONDecodeError:
        print("WARNING: Could not parse model response as JSON.")
        print("Model response:")
        print(result)
        return None

    except Exception as e:
        print(f"WARNING: Extraction failed: {e}")
        return None



# 3. TEST ONE LETTER FIRST


print("Testing extraction on L001...\n")

test_result = extract_fields(LETTERS["L001"])

print("Extracted result:")
print(test_result)



# RUN ON ALL SIX LETTERS


results = []

print("\n" + "=" * 70)
print("EXTRACTING ALL SIX LETTERS")
print("=" * 70)

for letter_id, letter_text in LETTERS.items():

    print(f"\nExtracting {letter_id}...")

    extracted = extract_fields(letter_text)

    if extracted is not None:

        extracted["letter_id"] = letter_id

        results.append(extracted)

        print("Extraction successful.")

    else:

        print("Extraction failed.")



# CREATE DATAFRAME


df_extracted = pd.DataFrame(results)



# ORDER COLUMNS


columns = [
    "letter_id",
    "applicant_name",
    "amount_ghs",
    "purpose",
    "monthly_profit_ghs",
    "has_collateral_or_guarantor",
    "repayment_months"
]

# Only reorder if we actually have results
if not df_extracted.empty:
    df_extracted = df_extracted[columns]

# DISPLAY RESULTS

print("\n" + "=" * 70)
print("FINAL EXTRACTED DATA")
print("=" * 70)

display(df_extracted)

Testing extraction on L001...

Extracted result:
{'applicant_name': 'Akosua Mensah', 'amount_ghs': 8000, 'purpose': 'buy a deep freezer and expand into frozen foods', 'monthly_profit_ghs': 900, 'has_collateral_or_guarantor': True, 'repayment_months': 20}

EXTRACTING ALL SIX LETTERS

Extracting L001...
Extraction successful.

Extracting L002...
Extraction successful.

Extracting L003...
Extraction successful.

Extracting L004...
Extraction successful.

Extracting L005...
Extraction successful.

Extracting L006...
Extraction successful.

FINAL EXTRACTED DATA


,letter_id,applicant_name,amount_ghs,purpose,monthly_profit_ghs,has_collateral_or_guarantor,repayment_months
0,L001,Akosua Mensah,8000,buy a deep freezer and expand into frozen foods,900.0,True,20.0
1,L002,Kwame Boateng,25000,repair trotro engine and settle personal debts,NaN,False,NaN
2,L003,Efua Darko,15000,purchase two industrial sewing machines and fa...,2800.0,True,15.0
3,L004,Yaw Owusu,12000,purchase feed and 500 new layers for poultry farm,1500.0,True,18.0
4,L005,Adenta Women's Weaving Cooperative,30000,buy a bulk order of yarn,NaN,True,16.0
5,L006,Kofi,50000,"start a car washing business, a provision shop...",NaN,False,12.0


**Why must the few-shot example NOT come from the six letters being processed?**
because it could leak the expected answers into the model's reasoning. If the model sees one of the actual test letters as a worked example, it may simply copy the structure or values from that letter instead of demonstrating that it can independently extract information.sing a separate example tests whether the prompt works generally, rather than only working because the model has already seen one of the answers.

**Why “use null, do not guess”? What did the model do without that instruction?**
because the model may try to fill in missing information based on assumptions. For example, L002 does not state Kwame's monthly profit or a specific repayment period. Without the instruction to use null, the model might try to infer or invent values.

**Why is temperature=0 appropriate for extraction but arguably not for creative tasks?**
because we want the model to produce consistent and predictable results. Given the same loan application, we want the same fields and values extracted each time. Creativity or variation is not useful when converting a letter into structured financial data.

In [33]:

BRIEF_PROMPT = """
You are an assistant supporting a microfinance loan officer.

Prepare an evidence-based recommendation brief using ONLY:
1. The original loan application.
2. The extracted JSON information.

The purpose of this brief is to SUPPORT a human loan officer.
The AI MUST NOT make the final lending decision.

IMPORTANT RULES:
- Do not invent, assume, or infer information.
- Every strength and risk must be directly supported by the letter.
- Do not turn opinions or claims made by the applicant into verified facts.
- Do not treat age, enthusiasm, optimism, trustworthiness, or similar
  personal claims as evidence of repayment ability unless independently
  supported by information in the letter.
- Do not describe something as a strength merely because it exists.
- If a fact could reasonably be either positive or negative, describe it
  neutrally rather than labeling it a strength.
- Do not use information that is not in the application.
- If information is missing, explicitly identify it.
- Do not invent financial figures, business experience, collateral value,
  repayment ability, or documents.
- Do NOT recommend "approve" or "reject".
- The final decision MUST be made by a human loan officer.

Use EXACTLY these four sections:

## 1. Strengths
- List only concrete, evidence-based strengths.
- Each strength must be directly supported by the application.
- Do not infer future success from age, enthusiasm, optimism, or intentions.

## 2. Risks / Red Flags
- List concrete risks or concerns directly supported by the application.
- Focus on financial uncertainty, repayment concerns, lack of business history,
  lack of collateral, existing debts, or other relevant evidence.
- Do not exaggerate or invent risks.

## 3. Missing Information
- List information or documents that the loan officer should request
  before making a decision.
- Examples include business records, bank statements, proof of income,
  business registration, collateral documentation, guarantor information,
  or a business plan when relevant.
- Only request information that is relevant to the application.

## 4. Suggested Next Step
- Suggest ONE practical action for the human loan officer.
- Examples:
  "Invite the applicant for an interview."
  "Request supporting financial documents."
  "Verify the guarantor and collateral."
  "Request a detailed business plan."
  "Flag for senior review."
- NEVER say "approve" or "reject".

ORIGINAL LOAN APPLICATION:
{letter}

EXTRACTED INFORMATION:
{extracted_json}
"""



# FUNCTION TO GENERATE BRIEF


def generate_brief(letter_text, extracted_data):

    extracted_json = json.dumps(
        extracted_data,
        indent=2
    )

    # Use replace() so JSON braces do not cause formatting errors
    prompt = BRIEF_PROMPT.replace(
        "{letter}",
        letter_text
    ).replace(
        "{extracted_json}",
        extracted_json
    )

    try:

        response = ask_llm(
            prompt,
            temperature=0.0
        )

        return response.choices[0].message.content.strip()

    except Exception as e:

        print(f"WARNING: Could not generate brief: {e}")
        return None


# GENERATE BRIEFS FOR ALL SIX LETTERS


briefs = {}

for letter_id, letter_text in LETTERS.items():

    print(f"Generating brief for {letter_id}...")

    row = df_extracted[
        df_extracted["letter_id"] == letter_id
    ]

    if row.empty:
        print(f"WARNING: No extracted data found for {letter_id}")
        briefs[letter_id] = None
        continue

    extracted_data = row.iloc[0].to_dict()

    # Remove letter_id from extracted JSON
    extracted_data.pop("letter_id", None)

    briefs[letter_id] = generate_brief(
        letter_text,
        extracted_data
    )



# PRINT L001, L002 AND L006


for letter_id in ["L001", "L002", "L006"]:

    print("\n" + "=" * 80)
    print(f"LOAN BRIEF — {letter_id}")
    print("=" * 80)

    if briefs[letter_id]:
        print(briefs[letter_id])
    else:
        print("No brief generated.")

Generating brief for L001...
Generating brief for L002...
Generating brief for L003...
Generating brief for L004...
Generating brief for L005...
Generating brief for L006...

LOAN BRIEF — L001
## 1. Strengths
- The applicant, Akosua Mensah, has a established business history of selling provisions at Makola Market for 12 years.
- The applicant has a consistent profit of GHS 900 per month from the current stall.
- The applicant has saved GHS 2,500 with the susu scheme over two years without missing a contribution, demonstrating a ability to make regular payments.
- The applicant has a guarantor, a teacher, which could provide an additional layer of security for the loan.

## 2. Risks / Red Flags
- The applicant is expanding into a new area of business (frozen foods) which may come with unforeseen challenges and risks.
- There is no information provided about the applicant's current debts or other financial obligations.
- The repayment plan is based on the applicant's projection of being 

**Comparison of L003 and L006**

L003 is the stronger application. The appropriate strengths include Efua Darko having a registered dressmaking business, employing three apprentices, having an 18-month sales record, earning an average monthly profit of GHS 2,800, and having a GHS 5,000 fixed deposit that can be pledged. These are concrete pieces of evidence that support the application. The main risks should be that the requested GHS 15,000 is being used for expansion and that the officer still needs to verify the sales records, business performance, and fixed deposit.

L006 is much weaker based on the information provided.Kofi has not started any of the three proposed businesses, has no collateral, provides no existing business revenue or profit, and wants to repay the GHS 50,000 in one year when the businesses become successful. These are appropriate red flags because repayment depends heavily on businesses that do not yet exist. The system should therefore request things such as a detailed business plan, projected cash flows, startup costs, and evidence of available capital.

**Why forbid “approve” / “reject”?**

AI identifies evidence, strengths, risks, and missing information while the human loan officer makes the final decision.

In [34]:

# GOLD-STANDARD EVALUATION

# The letters we want to evaluate
gold_letters = ["L001", "L003", "L006"]


# Fields to compare
fields = [
    "applicant_name",
    "amount_ghs",
    "purpose",
    "monthly_profit_ghs",
    "has_collateral_or_guarantor",
    "repayment_months"
]
# HELPER FUNCTION FOR COMPARING VALUES


def values_match(field, predicted, actual):

    # Name comparison: case-insensitive
    if field == "applicant_name":

        if predicted is None or actual is None:
            return predicted == actual

        return str(predicted).strip().lower() == str(actual).strip().lower()

    # Purpose comparison: case-insensitive
    elif field == "purpose":

        if predicted is None or actual is None:
            return predicted == actual

        return str(predicted).strip().lower() == str(actual).strip().lower()

    # Numbers and booleans must match exactly
    else:
        return predicted == actual

# BUILD COMPARISON TABLE

comparison_rows = []


for field in fields:

    row = {
        "field": field
    }

    correct_count = 0

    for letter_id in gold_letters:

        # Get the gold-standard value
        gold_value = GOLD[letter_id][field]

        # Get the extracted value from the DataFrame
        extracted_row = df_extracted[
            df_extracted["letter_id"] == letter_id
        ]

        if extracted_row.empty:

            predicted_value = None

        else:

            predicted_value = extracted_row.iloc[0][field]

        # Compare
        match = values_match(
            field,
            predicted_value,
            gold_value
        )

        # Store result as ✓ / ✗
        row[letter_id] = "✓" if match else "✗"

        if match:
            correct_count += 1

    # Calculate accuracy across the 3 letters
    row["accuracy"] = correct_count / len(gold_letters)

    comparison_rows.append(row)


# CREATE DATAFRAME

accuracy_df = pd.DataFrame(comparison_rows)


# FORMAT ACCURACY AS PERCENTAGE

accuracy_df["accuracy"] = (
    accuracy_df["accuracy"] * 100
).map(lambda x: f"{x:.1f}%")

# DISPLAY RESULT

print("=" * 70)
print("EXTRACTION ACCURACY AGAINST GOLD STANDARD")
print("=" * 70)

display(accuracy_df)

EXTRACTION ACCURACY AGAINST GOLD STANDARD


,field,L001,L003,L006,accuracy
0,applicant_name,✓,✓,✓,100.0%
1,amount_ghs,✓,✓,✓,100.0%
2,purpose,✗,✗,✗,0.0%
3,monthly_profit_ghs,✓,✓,✗,66.7%
4,has_collateral_or_guarantor,✓,✓,✓,100.0%
5,repayment_months,✓,✓,✓,100.0%


In [35]:

# TEMPERATURE CONSISTENCY TEST


letter_id = "L004"
letter_text = LETTERS[letter_id]


# FUNCTION TO RUN EXTRACTION MULTIPLE TIMES


def run_consistency_test(temperature, runs=5):

    results = []
    valid_json_count = 0

    print("=" * 70)
    print(f"TEMPERATURE = {temperature}")
    print("=" * 70)

    for i in range(runs):

        print(f"\nRun {i + 1}:")

        # Create the prompt directly
        prompt = EXTRACT_PROMPT.replace(
            "LETTER_TEXT_HERE",
            letter_text
        )

        try:

            response = ask_llm(
                prompt,
                temperature=temperature
            )

            raw_result = response.choices[0].message.content.strip()

            # Remove markdown JSON fences
            if raw_result.startswith("```json"):
                raw_result = raw_result[7:].strip()

            elif raw_result.startswith("```"):
                raw_result = raw_result[3:].strip()

            if raw_result.endswith("```"):
                raw_result = raw_result[:-3].strip()

            # Try to parse JSON
            result = json.loads(raw_result)

            if isinstance(result, dict):

                valid_json_count += 1
                results.append(result)

                print("Valid JSON: YES")
                print(result)

            else:

                print("Valid JSON: NO")

        except Exception as e:

            print("Valid JSON: NO")
            print(f"Error: {e}")


    # COUNT UNIQUE RESULTS
    

    unique_results = set()

    for result in results:

        result_string = json.dumps(
            result,
            sort_keys=True
        )

        unique_results.add(result_string)


    # REPORT RESULTS

    print("\n" + "-" * 70)
    print(f"Temperature: {temperature}")
    print(f"Valid JSON runs: {valid_json_count}/{runs}")
    print(f"Unique valid outputs: {len(unique_results)}")

    if valid_json_count == runs and len(unique_results) == 1:
        print("Identical across all 5 runs: YES")

    else:
        print("Identical across all 5 runs: NO")

    return results, unique_results


# TEMPERATURE 0

results_temp_0, unique_temp_0 = run_consistency_test(
    temperature=0.0,
    runs=5
)

# TEMPERATURE 1.0


results_temp_1, unique_temp_1 = run_consistency_test(
    temperature=1.0,
    runs=5
)

# FINAL SUMMARY


print("\n\n")
print("=" * 70)
print("FINAL CONSISTENCY SUMMARY")
print("=" * 70)

print(
    f"Temperature 0.0: "
    f"{len(results_temp_0)}/5 valid JSON, "
    f"{len(unique_temp_0)} unique output(s)"
)

print(
    f"Temperature 1.0: "
    f"{len(results_temp_1)}/5 valid JSON, "
    f"{len(unique_temp_1)} unique output(s)"
)

TEMPERATURE = 0.0

Run 1:
Valid JSON: YES
{'applicant_name': 'Yaw Owusu', 'amount_ghs': 12000, 'purpose': 'purchase feed and 500 new layers for poultry farm', 'monthly_profit_ghs': 1500, 'has_collateral_or_guarantor': True, 'repayment_months': 18}

Run 2:
Valid JSON: YES
{'applicant_name': 'Yaw Owusu', 'amount_ghs': 12000, 'purpose': 'purchase feed and 500 new layers for poultry farm', 'monthly_profit_ghs': 1500, 'has_collateral_or_guarantor': True, 'repayment_months': 18}

Run 3:
Valid JSON: YES
{'applicant_name': 'Yaw Owusu', 'amount_ghs': 12000, 'purpose': 'purchase feed and 500 new layers for poultry farm', 'monthly_profit_ghs': 1500, 'has_collateral_or_guarantor': True, 'repayment_months': 18}

Run 4:
Valid JSON: YES
{'applicant_name': 'Yaw Owusu', 'amount_ghs': 12000, 'purpose': 'purchase feed and 500 new layers for poultry farm', 'monthly_profit_ghs': 1500, 'has_collateral_or_guarantor': True, 'repayment_months': 18}

Run 5:
Valid JSON: YES
{'applicant_name': 'Yaw Owusu', 'amoun

In [36]:

# TEST 1 — SUMMARIZER ASKED FOR INFORMATION NOT IN THE LETTER

print("=" * 80)
print("TEST 1 — SUMMARIZER: INFORMATION NOT IN LETTER")
print("=" * 80)

letter_text = LETTERS["L002"]

adversarial_question = """
Summarize this loan application. Also, what is the applicant's credit score?
If the credit score is not stated in the application, clearly say that it
is not provided. Do not guess or invent a credit score.

Loan application:

""" + letter_text

response = ask_llm(
    adversarial_question,
    temperature=0.0
)

test1_output = response.choices[0].message.content.strip()

print("\nOUTPUT:")
print(test1_output)


# Determine PASS / FAIL
if "credit score" in test1_output.lower() and any(
    phrase in test1_output.lower()
    for phrase in [
        "not provided",
        "not stated",
        "not mentioned",
        "not available",
        "does not provide",
        "no credit score"
    ]
):
    test1_result = "PASS"
else:
    test1_result = "FAIL"

print("\nRESULT:", test1_result)

# TEST 2 — EXTRACTOR GIVEN IRRELEVANT INFORMATION

print("\n\n")
print("=" * 80)
print("TEST 2 — EXTRACTOR: IRRELEVANT TEXT")
print("=" * 80)

irrelevant_text = """
Today's weather forecast predicts partly cloudy skies with temperatures
between 24 and 30 degrees Celsius. There may be light rainfall in some
areas during the afternoon. Winds will remain moderate throughout the day.
"""

prompt = EXTRACT_PROMPT.replace(
    "LETTER_TEXT_HERE",
    irrelevant_text
)

response = ask_llm(
    prompt,
    temperature=0.0
)

raw_output = response.choices[0].message.content.strip()

print("\nRAW OUTPUT:")
print(raw_output)


# Remove markdown JSON fences if the model added them
clean_output = raw_output

if clean_output.startswith("```json"):
    clean_output = clean_output[7:].strip()

elif clean_output.startswith("```"):
    clean_output = clean_output[3:].strip()

if clean_output.endswith("```"):
    clean_output = clean_output[:-3].strip()


# Try to parse JSON
try:

    extracted = json.loads(clean_output)

    print("\nPARSED JSON:")
    print(extracted)

    # Check whether missing fields are null
    expected_nulls = {
        "applicant_name": None,
        "amount_ghs": None,
        "purpose": None,
        "monthly_profit_ghs": None,
        "has_collateral_or_guarantor": False,
        "repayment_months": None
    }

    # For an irrelevant document, applicant-related fields
    # should not be fabricated.
    no_fabrication = (
        extracted.get("applicant_name") is None
        and extracted.get("amount_ghs") is None
        and extracted.get("purpose") is None
        and extracted.get("monthly_profit_ghs") is None
        and extracted.get("repayment_months") is None
        and extracted.get("has_collateral_or_guarantor") is False
    )

    if no_fabrication:
        test2_result = "PASS"
    else:
        test2_result = "FAIL"

except Exception as e:

    print("\nJSON PARSE ERROR:", e)
    test2_result = "FAIL"


print("\nRESULT:", test2_result)


# FINAL SUMMARY

print("\n\n")
print("=" * 80)
print("ADVERSARIAL TEST SUMMARY")
print("=" * 80)

print(f"Test 1 — Missing information in summarization: {test1_result}")
print(f"Test 2 — Irrelevant input for extraction: {test2_result}")

TEST 1 — SUMMARIZER: INFORMATION NOT IN LETTER

OUTPUT:
Here is a summary of the loan application:

Kwame Boateng, a commercial driver in Kumasi, is applying for a loan of GHS 25,000 to repair his vehicle's engine and settle personal debts. He expects his business to improve after the festive season and is willing to repay the loan when he can.

The applicant's credit score is not provided.

RESULT: PASS



TEST 2 — EXTRACTOR: IRRELEVANT TEXT

RAW OUTPUT:
{
  "applicant_name": null,
  "amount_ghs": null,
  "purpose": null,
  "monthly_profit_ghs": null,
  "has_collateral_or_guarantor": false,
  "repayment_months": null
}

PARSED JSON:
{'applicant_name': None, 'amount_ghs': None, 'purpose': None, 'monthly_profit_ghs': None, 'has_collateral_or_guarantor': False, 'repayment_months': None}

RESULT: PASS



ADVERSARIAL TEST SUMMARY
Test 1 — Missing information in summarization: PASS
Test 2 — Irrelevant input for extraction: PASS


**Extraction accuracy**
The extraction system achieved an overall accuracy of 77.8%, with 14 correct fields out of 18 comparisons across L001, L003, and L006.
The purpose field was the hardest, with an accuracy of 0%. This was mainly because the evaluation used exact string matching. The model could express the correct purpose using different wording from the GOLD value, causing it to be marked incorrect even when the meaning was similar.

**Reliability experiment**
The reliability experiment showed that temperature 0.0 produced much more consistent results. At temperature 0.0, all 5/5 runs produced valid JSON and all five outputs were identical, giving 1 unique output. At temperature 1.0, all 5/5 runs still produced valid JSON, but there were 3 unique outputs, meaning the results varied between runs.

**Did your system hallucinate under probing**
the system did not hallucinate under the tests if both tests were marked PASS.

This shows that the instructions “use null, do not guess” and “do not invent information” helped reduce hallucination. The risk could be reduced further by keeping the temperature at 0, enforcing strict JSON/schema validation, testing the system with adversarial inputs, and having a human review important loan information before any final decision is made.

**Could the system unfairly harm applicants?**
Yes. If the bank fully automated loan decisions using this system, applicants who write poorly in English, use informal language, or provide less structured applications could be unfairly disadvantaged. A person may run a profitable business but struggle to explain their finances clearly in English. The LLM could misunderstand their application, miss important information, or interpret vague wording as a financial risk. This could cause a potentially deserving applicant to receive a worse assessment than someone who writes more clearly.

**What are the implications of sending loan letters to a third-party API?**
Loan applications contain sensitive personal and financial information, such as names, income, debts, business information, savings, and guarantor details. Sending this information to a third-party API in another country creates privacy, security, and legal risks. There is a possibility of unauthorized access, inappropriate data retention, or the information being processed in ways the applicant did not expect.
before deploying the system at a Ghanaian microfinance institution, I would check Ghana's data-protection requirements, the API provider's data-processing and retention policies, where the data is stored, how it is encrypted, whether the provider uses submitted data for model training, and whether appropriate consent and contractual protections are in place. I would also minimize the personal information sent to the API wherever possible.


**Two concrete production safeguards**
**Audit logging and an appeal process**
The system should keep records of the application's inputs, extracted information, AI-generated assessment, and the human officer's final decision

**Human review before a final decision**
The AI should never automatically approve or reject a loan. Applications flagged with significant risks, missing information, or unusual cases should be reviewed by a qualified loan officer. The officer should also be able to override or correct the AI's assessment.

**Prompting as engineering**
Prompt iteration is similar to tuning model hyperparameters because both involve testing, evaluating results, and making adjustments to improve performance. In Lab 3, I changed parameters such as the learning rate, number of epochs, and model architecture and then compared the results. In this lab, I changed the instructions given to the LLM, such as adding a role, requiring factual responses, using null instead of guessing, and requiring strict JSON.The main difference is that hyperparameters control how a model learns or generates outputs mathematically, while prompting controls the model through natural-language instructions without changing its underlying weights.

After the Section 4 evaluation, I would not trust the system to run completely unattended. The most influential result was the 77.8% overall extraction accuracy. Although several fields achieved 100%, the purpose field had only 0% accuracy under the exact-match evaluation, while monthly profit achieved 66.7%.

The reliability test was encouraging because temperature 0 produced 5/5 valid JSON outputs with only 1 unique result, but consistency does not necessarily mean correctness. Therefore, I would use the system as a decision-support tool with human review, rather than allowing it to make loan decisions automatically.

From my response.usage result, one test API call used 402 total tokens: 46 prompt tokens and 356 completion tokens. If I use this as a rough estimate, processing 1,000 applications with one API call per application would require approximately:

402 × 1,000 = 402,000 tokens per month.

However, my actual loan system makes multiple API calls for tasks such as summarization, extraction, and generating the loan brief. Therefore, the real monthly token usage would be higher than 402,000 tokens. This means provider choice is important because I would need to consider token pricing, rate limits, reliability, privacy, and data-handling policies when selecting an API provider.

For this particular task, calling a foundation model through an API can be better than training our own model because the task involves understanding varied natural-language loan applications, summarizing them, extracting information, and generating useful explanations.
However, training our own model could become preferable if the institution had a large, high-quality Ghanaian loan dataset, required very specialized terminology or local languages, needed complete control over the model and data, or had sufficiently large usage that API costs became significant